relayer/chains/cosmos/message_handlers.go

```diff
@@ -0,0 +1,187 @@
+	if !ccp.isRelayedChannel(channelKey) {
+		return false
+	}
+	if _, ok := foundMessages[channelKey]; !ok {
+		return true
+	}
+	if _, ok := foundMessages[channelKey][message]; !ok {
+		return true
+	}
+	for sequence := range foundMessages[channelKey][message] {
+		if sequence == packetInfo.packet.Sequence {
+			// already have this sequence number
+			// there can be multiple MsgRecvPacket, MsgAcknowlegement, MsgTimeout, and MsgTimeoutOnClose for the same packet
+			// from different relayers.
+			return false
+		}
+	}
+
+	return true
+}
+
+// retainPacketMessage assumes the packet is applicable to the channels for a path processor that is subscribed to this chain processor.
+// It creates cache path if it doesn't exist, then caches message.
+func retainPacketMessage(message string, packetInfo *packetInfo, foundMessages processor.ChannelMessageCache, channelKey processor.ChannelKey, ibcMsg provider.RelayerMessage) {
+	if _, ok := foundMessages[channelKey]; !ok {
+		foundMessages[channelKey] = make(processor.MessageCache)
+	}
+	if _, ok := foundMessages[channelKey][message]; !ok {
+		foundMessages[channelKey][message] = make(processor.SequenceCache)
+	}
+	foundMessages[channelKey][message][packetInfo.packet.Sequence] = ibcMsg
+}
+
+// BEGIN packet msg handlers
+func (ccp *CosmosChainProcessor) handleMsgTransfer(p MsgHandlerParams) bool {
+	packetInfo := p.messageInfo.(*packetInfo)
+	// source chain processor will call this handler
+	// source channel used as key because MsgTransfer is sent to source chain
+	channelKey := packetInfo.channelKey()
+	if !ccp.isPacketApplicable(processor.MsgTransfer, packetInfo, p.foundMessages, channelKey) {
+		return false
+	}
+	// Construct the start of the MsgRecvPacket for the counterparty chain.
+	// PathProcessor will determine if this is needed.
+	// For example, if a MsgRecvPacket is not detected on the counterparty chain, and
+	// a MsgAcknowledgement, MsgTimeout, or MsgTimeout is not detected yet on this chain,
+	// and the packet timeout has not expired, a MsgRecvPacket will be sent to the counterparty chain
+	// using this information with the packet commitment proof from this chain added.
+	retainPacketMessage(processor.MsgTransfer, packetInfo, p.foundMessages, channelKey, cosmos.NewCosmosMessage(&chantypes.MsgRecvPacket{
+		Packet: chantypes.Packet{
+			Sequence:           packetInfo.packet.Sequence,
+			SourcePort:         packetInfo.packet.SourcePort,
+			SourceChannel:      packetInfo.packet.SourceChannel,
+			DestinationPort:    packetInfo.packet.DestinationPort,
+			DestinationChannel: packetInfo.packet.DestinationChannel,
+			Data:               packetInfo.packet.Data,
+			TimeoutHeight:      packetInfo.packet.TimeoutHeight,
+			TimeoutTimestamp:   packetInfo.packet.TimeoutTimestamp,
+		},
+	}))
+	ccp.logPacketMessage("MsgTransfer", packetInfo,
+		zap.String("timeout_height", fmt.Sprintf("%d-%d", packetInfo.packet.TimeoutHeight.RevisionNumber, packetInfo.packet.TimeoutHeight.RevisionHeight)),
```

Comment:
Not sure if it's necessary to format the timeout height this way, vs. providing two `zap.Int` (or appropriate type) calls.

The `fmt.Sprintf` call will cause an allocation on every call even if the log level is ignored. You can probably use `zap.Stringer("timeout_height", ...TimeoutHeight)` since the TimeoutHeight type has a String method, which will then be lazily evaluated, and is probably already in the format you're looking for.

Reply:
done